### DISCUSSION TITLE
Grammar for Shopify Liquid!

### DISCUSSION THREAD

**Discussion by S0AndS0** [pest-parser/pest#1021] [Show and tell]
I've been learning the Pest, and some Rust with it too, and thought I might share a project I published today!

**TLDR:** Liquid is a templating language, often used for generating static web-sites.

- [Crates](https://crates.io/crates/liquid-grammar-pest): `https://crates.io/crates/liquid-grammar-pest`
- [Source](https://github.com/rust-utilities/liquid-grammar-pest): `https://github.com/rust-utilities/liquid-grammar-pest`
- [Grammar](https://github.com/rust-utilities/liquid-grammar-pest/blob/main/src/grammars/liquid.pest): `https://github.com/rust-utilities/liquid-grammar-pest/blob/main/src/grammars/liquid.pest`
- [Spec](https://shopify.github.io/liquid/): `https://shopify.github.io/liquid/`

______

There are a number of _flavors_ of Liquid but, for reasons, I chose to target the core Shopify variant first; specification can be found on their [GitHub Pages](https://shopify.github.io/liquid/) site.

> Note; attempts have been made to ensure the base grammar allows for consuming crates to implement additional features from other Liquid flavors.  But I've no doubt something(s) may have been missed X-D

## Example -- Liquid

```liquid
{% assign name = include.name | default: 'WaT' %}

{%- for item in site.collection[name] %}
  {%- if item.published %}
    <p>Item: {{ item }}</p>
  {%- endif %}
{%- endfor %}
```

> Note; the prefixed hyphens (`{%-`) are for [Whitespace](https://shopify.github.io/liquid/basics/whitespace/) control, and capturing the variants of stripping before, after, or both kinda bloated the Pest definitions.

## Example -- Parser result

```
- document
  - liquid__tag__assign
    - liquid__tag__open__plain: "{%"
    - liquid__code__assign
      - liquid__key_word__assign: "assign"
      - liquid__code__assign__key > object__word: "name"
      - liquid__code__assign__value > liquid__type > object
        - object__word: "include"
        - object__accessor > object__property > object__word: "name"
      - liquid__filter__list > liquid__filter__default
        - liquid__key_word__filter__default: "default"
        - colon: ":"
        - liquid__filter__default__value > liquid__type > string > string__quote__single
          - quote__single: "'"
          - string__quote__single__content: "WaT"
          - quote__single: "'"
    - liquid__tag__close__plain: "%}"
  - content__block: "\n\n"
  - liquid__block__iteration__for > liquid__block__iteration__for__strip_leading
    - liquid__block_subsection__iteration__for__strip_leading__start
      - liquid__tag__open__strip: "{%-"
      - liquid__key_word__iteration__for: "for"
      - object__word: "item"
      - liquid__code__iteration__target > object
        - object__word: "site"
        - object__accessor
          - object__property > object__word: "collection"
          - object__accessor > object__key
            - bracket__square__open: "["
            - object > object__word: "name"
            - bracket__square__close: "]"
      - liquid__tag__close__plain: "%}"
      - content__block: "\n  "
      - liquid__block__control_flow__branch__within_iteration > liquid__block__control_flow__branch__within_iteration__strip_leading
        - liquid__block_subsection__control_flow__branch__within_iteration__strip_leading__start
          - liquid__tag__open__strip: "{%-"
          - liquid__key_word__control_flow__branch: "if"
          - liquid__code__control_flow__branch__comparison__expression > liquid__code__control_flow__branch__comparison__operator__word > liquid__type > object
            - object__word: "item"
            - object__accessor > object__property > object__word: "published"
          - liquid__tag__close__plain: "%}"
          - content__block: "\n    <p>Item: {{ item }}</p>\n  "
        - liquid__tag__strip_leading__end
          - liquid__tag__open__strip: "{%-"
          - liquid__code__end: "endif"
          - liquid__tag__close__plain: "%}"
      - content__block: "\n"
    - liquid__tag__strip_leading__end
      - liquid__tag__open__strip: "{%-"
      - liquid__code__end: "endfor"
      - liquid__tag__close__plain: "%}"
  - content__block: "\n"
  - EOI: ""
```

At this time I'd be super grateful for some bug reports, and any suggestions that'd make the `src/grammars/liquid.pest` easier for consuming crates to utilize.

**Comment by tomtau** [comment]
Nice! Feel free to open a PR to add it to this list here: https://github.com/pest-parser/awesome-pest
